pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/ForwardIndexHandler.java

```diff
@@ -0,0 +1,273 @@
+ * majority of the usecases are in versions >= 3.0 and this avoids adding tech debt. The currently supported
+ * operations are:
+ * 1. Change compression on raw SV columns.
+ *
+ *  TODO: Add support for the following:
+ *  1. Change compression for raw MV columns
+ *  2. Enable dictionary
+ *  3. Disable dictionary
+ */
+public class ForwardIndexHandler implements IndexHandler {
+  private static final Logger LOGGER = LoggerFactory.getLogger(ForwardIndexHandler.class);
+
+  private final SegmentMetadata _segmentMetadata;
+  IndexLoadingConfig _indexLoadingConfig;
+
+  protected enum Operation {
+    // TODO: Add other operations like ENABLE_DICTIONARY, DISABLE_DICTIONARY.
+    CHANGE_RAW_INDEX_COMPRESSION_TYPE,
+  }
+
+  public ForwardIndexHandler(SegmentMetadata segmentMetadata, IndexLoadingConfig indexLoadingConfig) {
+    _segmentMetadata = segmentMetadata;
+    _indexLoadingConfig = indexLoadingConfig;
+  }
+
+  @Override
+  public boolean needUpdateIndices(SegmentDirectory.Reader segmentReader)
+      throws Exception {
+    Map<String, Operation> columnOperationMap = computeOperation(segmentReader);
+    return !columnOperationMap.isEmpty();
+  }
+
+  @Override
+  public void updateIndices(SegmentDirectory.Writer segmentWriter, IndexCreatorProvider indexCreatorProvider)
+      throws Exception {
+    Map<String, Operation> columnOperationMap = computeOperation(segmentWriter);
+    if (columnOperationMap.isEmpty()) {
+      return;
+    }
+
+    for (Map.Entry<String, Operation> entry : columnOperationMap.entrySet()) {
+      String column = entry.getKey();
+      Operation operation = entry.getValue();
+
+      switch (operation) {
+        case CHANGE_RAW_INDEX_COMPRESSION_TYPE:
+          rewriteRawForwardIndex(column, segmentWriter, indexCreatorProvider);
+          break;
+        // TODO: Add other operations here.
+        default:
+          throw new IllegalStateException("Unsupported operation for column " + column);
+      }
+    }
+  }
+
+  @VisibleForTesting
+  Map<String, Operation> computeOperation(SegmentDirectory.Reader segmentReader)
+      throws Exception {
+    Map<String, Operation> columnOperationMap = new HashMap<>();
+
+    // Does not work for segment versions < V3
+    if (_segmentMetadata.getVersion().compareTo(SegmentVersion.v3) < 0) {
+      return columnOperationMap;
+    }
+
+    // From existing column config.
+    Set<String> existingAllColumns = _segmentMetadata.getAllColumns();
+    Set<String> existingDictColumns =
+        segmentReader.toSegmentDirectory().getColumnsWithIndex(ColumnIndexType.DICTIONARY);
+    Set<String> existingNoDictColumns = new HashSet<>();
+    for (String column : existingAllColumns) {
+      if (!existingDictColumns.contains(column)) {
+        existingNoDictColumns.add(column);
+      }
+    }
+
+    // From new column config.
+    Set<String> newNoDictColumns = _indexLoadingConfig.getNoDictionaryColumns();
+
+    for (String column : existingAllColumns) {
```

Comment:
We can just iterate over `existingNoDictColumns` imo

We are anyway not supporting (in this PR) enabling / disabling dictionary on an existing column on the reload path. So, our assumption is that `existingNoDictColumns` and `newNoDictColumns` is exactly the same. 

If `newNoDictColumns` has more columns, it could mean any of the 2 things:

- User added a new column to schema and marked as noDict
- User put an existing dict column in the schema as noDict

(1) is already supported in `BaseDefaultColumnHandler` and should already be handled before reaching here. (2) is not supported yet

So this code should ignore (which you are already doing at the if check in line 129)

If `newNoDictColumns` has fewer columns, it could only mean the following

- User marked an existing nodict column in the schema as dict

This is also not supported and we can ignore.

Reply:
The reason for choosing to iterate over existingAllColumns is so that we can support enable and disable dictionary in the subsequent PRs. Would prefer keeping it this way so that we don't have to change when we add enable/disable dictionary support. Thoughts?

If we see my first commit, I initially had the block to check if dictionary is enabled or disabled with a TODO. Removed it in the subsequent commits to keep this PR confined.